File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
from collections import deque
from typing import List

class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)
        sorted_nums = sorted(nums)
        mpp1 = {} 
        mpp2 = {} 

        grp = 0
        mpp1[grp] = deque([sorted_nums[0]])

        for i in range(1, n):
            if sorted_nums[i] - sorted_nums[i - 1] > limit:
                grp += 1
            if grp not in mpp1:
                mpp1[grp] = deque()
            mpp1[grp].append(sorted_nums[i])
            mpp2[sorted_nums[i]] = grp
        
        for grp, dq in mpp1.items():
            for num in dq:
                mpp2[num] = grp

        ans = []
        for num in nums:
            resp_grp = mpp2[num]
            ans.append(mpp1[resp_grp].popleft())
        return ans

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<int> sortedNums = nums;
        sort(sortedNums.begin(),sortedNums.end());
        unordered_map<int,deque<int>> mpp1;
        unordered_map<int,int> mpp2;
        int grp = 0;
        mpp1[grp].push_back(sortedNums[0]);
        for(int i=1; i<n; i++){
            if(sortedNums[i] - sortedNums[i-1] > limit){
                grp++;
            }
            mpp1[grp].push_back(sortedNums[i]);
            mpp2[sortedNums[i]] = grp;
        }
        vector<int> ans;
        for(int i=0; i<n; i++){
            int respGrp = mpp2[nums[i]];
            ans.push_back(mpp1[respGrp].front());
            mpp1[respGrp].pop_front();
        }
        return ans;
    }
};